modules/data.atmosphere/R/met_temporal_downscale.Gaussian_ensemble.R

```diff
@@ -0,0 +1,398 @@
+                                    POSIXct.out = TRUE)
+      sunset <- maptools::sunriset(lon.lat, sequence, direction = "sunset", 
+                                   POSIXct.out = TRUE)
+      solar_noon <- maptools::solarnoon(lon.lat, sequence, POSIXct.out = TRUE)
+      
+      data.frame(date = as.Date(sunrise$time), sunrise = as.numeric(format(sunrise$time, "%H%M")), 
+                 solarnoon = as.numeric(format(solar_noon$time, "%H%M")), 
+                 sunset = as.numeric(format(sunset$time, "%H%M")),
+                 day_length = as.numeric(sunset$time - sunrise$time))
+    }
+    
+    swsource <- source$surface_downwelling_shortwave_flux_in_air
+    swdn <- vector()
+    
+    # The sine swdn_method produces an hourly sine wave of
+    if (swdn_method == "sine") {
+      
+      eph <- ephemeris(lat_train, lon_train, date = paste0(eph_year, "-01-01 00:00:00"), 
+                       span = sp, tz = "UTC")
+      day_len <- eph$day_length
+      
+      # Need to have average daily values for this swdn_method, so this upscales the
+      # source data to daily resolution if needed
+      daily_row <- nrow(source)
+      daily_step <- daily_row/sp
+      daily.swdn <- vector()
+      for (x in 1:sp) {
+        daily.swdn[x] <- mean(swsource[(x * daily_step - daily_step + 1):(x * daily_step)])
+      }
+      
+      # creating the sine wave
+      for (i in seq_along(daily.swdn)) {
+        t <- seq(from = pi/day_len[i], to = pi, by = pi/day_len[i])
+        wav <- ((daily.swdn[i] * (24/day_len[i]))/0.637) * sin(t)
+        
+        # swdn = 0 without sunlight
+        srs <- eph$sunrise
+        hr <- substr(srs[i], 1, 2)
+        hr <- as.numeric(hr)
+        hr <- hr + utc_diff
+        
+        l <- vector()
+        for (n in seq_along(1:hr)) {
+          l[n] <- 0
+        }
+        for (n in seq_along(wav)) {
+          l[n + hr] <- wav[n]
+        }
+        for (n in seq_along(1:(24 - (length(wav) + hr)))) {
+          l[n + hr + length(wav)] <- 0
+        }
+        
+        swdn <- append(swdn, l)
+      }
+      
+      swflux <- vector()
+      sw_step <- length(swdn)/reso_len
+      for (x in 1:reso_len) {
+        swflux[x] <- mean(swdn[(x * sw_step - sw_step + 1):(x * sw_step)])
+      }
+      swflux[swflux < 0] <- 0
+    }
+    
+    # The spline swdn_method uses spline interpolation to connect existing values and
+    # downscale
+    if (swdn_method == "spline") {
+      tem.met <- vector()
+      for (x in seq(from = 0, to = nrow(train), by = div)) {
+        tem.met[x] <- swsource[x/div]
+      }
+      
+      swflux <- vector()
+      swflux <- zoo::na.spline(tem.met)
+      swflux[swflux < 0] <- 0
+    }
+    
+    # The Waichler swdn_method doesn't need averaged SW train values, it sources SW
+    # downwelling flux based on Tmax-Tmin and Precipitation Reference is Waichler and
+    # Wigtosa 2003. Our no-precip coefficient is 2 instead of 1 (fits the data
+    # better)
```

Comment:
Only tested for Cheas sites, I'll have to test for others. The reason I just went with it was because using 1 for no precip factor didn't even account for half of the actual radiation, but I see your concern, perhaps it would be regionally dependent